### PR TITLE
disallow restricted XMLHttpRequest header prefixes

### PR DESCRIPTION
**Pull request by Chandler**
Servo currently filters on restricted headers but not restricted header prefixes ("sec-" and "proxy-")
https://www.w3.org/TR/XMLHttpRequest/#the-setrequestheader-method

This fixes two tests.

Not sure if I have the right syntax for that if block on xmlhttprequest.rs:393

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9376)

### CODE REVIEW

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -376,19 +376,25 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
...
-                    "connection" | "content-length" |
-                    "cookie" | "cookie2" | "date" |"dnt" |
-                    "expect" | "host" | "keep-alive" | "origin" |
-                    "referer" | "te" | "trailer" | "transfer-encoding" |
-                    "upgrade" | "user-agent" | "via" => {
-                        return Ok(()); // Step 5
-                    },
-                    _ => s
+                // Disallowed headers and header prefixes:
+                // https://www.w3.org/TR/XMLHttpRequest/#the-setrequestheader-method
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: We really shouldn't be using W3C spec links.

Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: Here's the WHATWG link: https://fetch.spec.whatwg.org/#forbidden-header-name

File: components/script/dom/xmlhttprequest.rs [hunk 2]
```diff
@@ -376,19 +376,25 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
...
+                let disallowedHeaders =
+                    ["accept-charset", "accept-encoding",
+                    "access-control-request-headers",
+                    "access-control-request-method",
+                    "connection", "content-length",
+                    "cookie", "cookie2", "date", "dnt",
+                    "expect", "host", "keep-alive", "origin",
+                    "referer", "te", "trailer", "transfer-encoding",
+                    "upgrade", "user-agent", "via"];
```
Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: Note that in the XHR WHATWG Spec, the "User-Agent" string is not a forbidden header name.

Comment (Ms2ger) [inline components/script/dom/xmlhttprequest.rs]: And we should make sure there's a test for this :)

**Comment by KiChjang** [comment]
Per @Ms2ger's comment, a test is still required to test the `setRequestHeader` API, particularly testing against "user-agent" and making sure it's an allowed header.

**Comment by Chandler** [comment]
@KiChjang @Ms2ger re: setRequestHeader tests

the web platform tests have a pretty good test for this here https://github.com/Chandler/servo/blob/add_disallowed_prefixes/tests/wpt/web-platform-tests/XMLHttpRequest/setrequestheader-header-forbidden.htm that are now passing with this change.

I assume we can't modify the wpt test to explicitly test that 'user-agent' is allowed, since it's forbidden in the w3c spec.

Should I add a servo only test outside of wpt to test this case? where would that go?

**Comment by Ms2ger** [comment]
No, you should edit the test under `tests/wpt/web-platform-tests`. The upstream location is maintained by W3C, but the WHATWG spec is canonical.

**Comment by Chandler** [comment]
Turns out, there was already test that explicitly tests if setting "User-Agent" was allowed.
https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/XMLHttpRequest/setrequestheader-header-allowed.htm

This test was failing because servo unconditionally modified the User-Agent to be the default, I've fixed it to only modify the User-Agent if it wasn't already set.

**Comment by Chandler** [comment]
bump

**Comment by Chandler** [comment]
can anyone take a look at this @jdm @KiChjang. Thanks much.

File: components/script/dom/xmlhttprequest.rs [hunk 3]
```diff
@@ -376,19 +376,25 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
...
+                    "access-control-request-method",
+                    "connection", "content-length",
+                    "cookie", "cookie2", "date", "dnt",
+                    "expect", "host", "keep-alive", "origin",
+                    "referer", "te", "trailer", "transfer-encoding",
+                    "upgrade", "via"];
+
+                let disallowedHeaderPrefixes = ["sec-", "proxy-"];
+
+                if disallowedHeaders.iter().any(|header| *header == s) |
```
Comment (jdm) [inline components/script/dom/xmlhttprequest.rs]: This should use || instead of |.

**Comment by jdm** [comment]
This is fine to merge with that small change made! Please amend the commit and force push the result.

File: components/net/http_loader.rs [hunk 4]
```diff
@@ -521,7 +521,16 @@ pub fn modify_request_headers(headers: &mut Headers,
...
-    headers.set(UserAgent(user_agent.to_owned()));
+
+    // If the user-agent has not already been set, then use the
+    // browser's default user-agent or the user-agent override
+    // from the command line. If the user-agent is set, don't
+    // modify it, as setting of the user-agent by the user is
+    // allowed.
+    // https://fetch.spec.whatwg.org/#concept-http-network-or-cache-fetch step 8
+    if !headers.has::<UserAgent>() {
+      headers.set(UserAgent(user_agent.to_owned()));
```
Comment (Ms2ger) [inline components/net/http_loader.rs]: Nit: this should be indented two more spaces.

**Comment by KiChjang** [comment]
@Chandler ping

**Comment by jdm** [comment]
Rebased in #9768.

**Pull request closed without merging**